go/pkg/bertymessenger/db.go

```diff
@@ -48,254 +62,709 @@ func dropAllTables(db *gorm.DB) error {
+func (d *dbWrapper) markInteractionAsAcknowledged(cid string) (bool, error) {
+	rowsAffected := int64(0)
+
+	if err := d.tx(func(db *dbWrapper) error {
+		tx := d.db.Model(&Interaction{}).Where(&Interaction{
+			CID: cid,
+		})
+
+		count := int64(0)
+
+		if err := tx.Count(&count).Error; err != nil {
+			return err
+		}
+
+		if count == 0 {
+			return errcode.ErrInvalidInput.Wrap(fmt.Errorf("interaction not found"))
+		}
+
+		tx = d.db.Model(&Interaction{}).Where(map[string]interface{}{"cid": cid, "acknowledged": false}).Update("acknowledged", true)
+		rowsAffected = tx.RowsAffected
+
+		return tx.Error
+	}); err != nil {
+		return false, err
+	}
+
+	return rowsAffected == 1, nil
+}
+
+func (d *dbWrapper) getAcknowledgementsCIDsForInteraction(cid string) ([]string, error) {
+	var cids []string
+
+	if err := d.db.Model(&Interaction{}).Where(&Interaction{
+		Type:      AppMessage_TypeAcknowledge,
+		TargetCID: cid,
+	}).Pluck("cid", &cids).Error; err != nil {
+		return nil, err
+	}
+
+	return cids, nil
+}
+
+func (d *dbWrapper) deleteInteractions(cids []string) error {
+	return d.db.Model(&Interaction{}).Delete(&Interaction{}, &cids).Error
+}
+
+func (d *dbWrapper) getDBInfo() (*SystemInfo_DB, error) {
+	var err, errs error
+	infos := &SystemInfo_DB{}
+
+	infos.Accounts, err = d.dbModelRowsCount(Account{})
+	errs = multierr.Append(errs, err)
+
+	infos.Contacts, err = d.dbModelRowsCount(Contact{})
+	errs = multierr.Append(errs, err)
+
+	infos.Interactions, err = d.dbModelRowsCount(Interaction{})
+	errs = multierr.Append(errs, err)
+
+	infos.Conversations, err = d.dbModelRowsCount(Conversation{})
+	errs = multierr.Append(errs, err)
+
+	infos.Members, err = d.dbModelRowsCount(Member{})
+	errs = multierr.Append(errs, err)
+
+	infos.Devices, err = d.dbModelRowsCount(Device{})
+	errs = multierr.Append(errs, err)
+
+	infos.ServiceTokens, err = d.dbModelRowsCount(ServiceToken{})
+	errs = multierr.Append(errs, err)
+
+	return infos, errs
+}
+
+func (d *dbWrapper) addDevice(devicePK string, memberPK string) (*Device, error) {
+	if err := d.tx(func(d *dbWrapper) error {
+		// Check if this device already exists for another member
+		{
+			count := int64(0)
+
+			if err := d.db.
+				Model(&Device{}).
+				Where("member_public_key != ? AND public_key == ?", memberPK, devicePK).
+				Count(&count).
+				Error; err != nil {
+				return err
+			}
+
+			if count > 0 {
+				return errcode.ErrDBWrite.Wrap(fmt.Errorf("the device already exists for another member"))
+			}
+		}
+
+		return d.db.
+			Clauses(clause.OnConflict{DoNothing: true}).
+			Create(&Device{
+				PublicKey:       devicePK,
+				MemberPublicKey: memberPK,
+			}).
+			Error
+	}); err != nil {
+		return nil, err
+	}
+
+	return d.getDeviceByPK(devicePK)
+}
+
+func (d *dbWrapper) updateContact(contact Contact) error {
+	if contact.PublicKey == "" {
+		return errcode.ErrInvalidInput.Wrap(fmt.Errorf("no public key specified"))
+	}
+
+	return d.tx(func(wrapper *dbWrapper) error {
```

Comment:
```suggestion
	return d.tx(func(tx *dbWrapper) error {
```

This variable is sometimes named `d`, sometimes `db`, sometimes, `wrapper`, what about just nameing it `tx` everywhere?